nixos/systemd: Implement reload triggers (#157329)
NixOS/nixpkgs: nixos/doc/manual/development/what-happens-during-a-system-switch.chapter.md

# What happens during a system switch? {#sec-switching-systems}

`nixos-rebuild switch` is one of the more common tasks under NixOS. This
chapter explains some of the things that are happening when doing this to make
it easier for new module developers to configure their units and to make it
easier for administrators who are curious what is happening and why.

`nixos-rebuild`, like many deployment solutions, calls `switch-to-configuration`
which resides in a NixOS system in `$out/bin/switch-to-configuration`. The
script is called with the action that is to be performed like `switch`, `test`,
`boot`. There is also the `dry-activate` action which does not really perform
the actions but rather prints what it would do if you called it with `test`.

> This feature can be used to check what services would be changed if a
> configuration is run.

Reviewer: ```suggestion
This feature can be used to check what services states would be changed if the
configuration was run.
```

`1`, `/nix/store` is synced to disk.

If the action is `switch` or `test`, the currently running system is inspected
and the actions to switch to the new system are calculated. This process takes
two data sources into account: `/etc/fstab` and the current systemd status.
Mounts and swaps are read from `/etc/fstab` and the corresponding actions are
generated. If a new mount is added, for example, the proper `.mount` unit is
marked to be started. The current systemd state is inspected and the difference
between the current system and the desired configuration is calculated and
actions are generated to get to this state. There are a lot of nuances that can
be controlled by the units which are explained here.

After calculating what should be done, the actions are carried out. The order
of actions is always the same: stop units, run activation script, see if the
activation script requested more units to restart, restart systemd if needed,
forget about the failed state of units, reload systemd, reload systemd user
instances, set up tmpfiles, reload units, restart units, start units, inspect
what changed during these actions. While this sounds like a lot (and it is),

> it's quite logical to go through all these steps which are explained here.

Reviewer 2: ~which are explained here.~ Does not add value

> Units are then stopped, the activation script is run, additional units that may

Reviewer 2: as far as I understand your documentation this section duplicates parts of the section above

Author: It's supposed to be a big picture overview of the process
